File: python/test/t_KissFFT_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()

myFFT = ot.KissFFT()
print("myFFT = ", myFFT)

# 1D case with deterministic data
size = 16
collection = ot.ComplexCollection(size)
for index in range(size):
    realPart = 0.1 * (index + 1.0) / size
    imagPart = 0.3 * (index + 1.0) / size
    collection[index] = realPart + 1j * imagPart

print("collection = ", collection)

transformedCollection = ot.ComplexCollection(myFFT.transform(collection))
print("FFT result = ", transformedCollection)

inverseCollection = ot.ComplexCollection(
    myFFT.inverseTransform(transformedCollection)
)
print("FFT back=", inverseCollection)

# Verify 1D roundtrip
for i in range(size):
    ott.assert_almost_equal(inverseCollection[i].real, collection[i].real, 1e-14, 1e-13)
    ott.assert_almost_equal(inverseCollection[i].imag, collection[i].imag, 1e-14, 1e-13)

# 2D case
N = 8
ot.RandomGenerator.SetSeed(0)
distribution = ot.Normal(N)
sample = distribution.getSample(2 * N)
rows = sample.getSize()
cols = sample.getDimension()
sampleFlat = ot.ComplexCollection(
    [complex(sample[i, j], 0.0) for j in range(cols) for i in range(rows)]
)

transformedSample = myFFT.transform(sampleFlat, [rows, cols])
print("2D FFT size = ", transformedSample.getSize())

inverseSample = myFFT.inverseTransform(transformedSample, [rows, cols])
print("2D FFT back size = ", inverseSample.getSize())

for i in range(rows * cols):
    ott.assert_almost_equal(inverseSample[i].real, sampleFlat[i].real, 1e-12, 1e-12)
    ott.assert_almost_equal(inverseSample[i].imag, sampleFlat[i].imag, 0.0, 1e-12)

# 3D case
ot.RandomGenerator.SetSeed(0)
elements = [ot.RandomGenerator.Generate() for i in range(N * N * N)]
tensorFlat = ot.ComplexCollection(
    [complex(elements[i + j * N + k * N * N], 0.0)
     for k in range(N) for j in range(N) for i in range(N)]
)

transformedTensor = myFFT.transform(tensorFlat, [N, N, N])
print("3D FFT size = ", transformedTensor.getSize())

inverseTensor = myFFT.inverseTransform(transformedTensor, [N, N, N])
print("3D FFT back size = ", inverseTensor.getSize())

for i in range(N * N * N):
    ott.assert_almost_equal(inverseTensor[i].real, tensorFlat[i].real, 1e-12, 1e-12)
    ott.assert_almost_equal(inverseTensor[i].imag, tensorFlat[i].imag, 0.0, 1e-12)

File: lib/test/t_KissFFT_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class KissFFT
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

typedef Collection<Complex> ComplexCollection;


int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    RandomGenerator::SetSeed(0);

    // for fft, the best implementation is given for N = 2^p
    const UnsignedInteger size = 16;

    // collection for test
    ComplexCollection collection(size);

    // Fill the data with artificial values

    // Create a complex gaussian sample
    for (UnsignedInteger index = 0; index < size; ++index)
    {
      const Scalar realPart = 0.1 * (1.0 + index) / size;
      const Scalar imagPart = 0.3 * (1.0 + index) / size;
      collection[index] = Complex(realPart, imagPart);
    }

    /* Instantiation of FFT class */
    KissFFT myFFT;
    fullprint << "myFFT = " << myFFT << std::endl;

    // 1D case
    const ComplexCollection transformedCollection(myFFT.transform(collection));
    const ComplexCollection inverseTransformedCollection(myFFT.inverseTransform(transformedCollection));
    for (UnsignedInteger i = 0; i < collection.getSize(); ++i)
      assert_almost_equal(inverseTransformedCollection[i], collection[i]);

    // 2D ND case
    const UnsignedInteger N = 8;
    Normal distribution(N);
    Sample sample(distribution.getSample(2 * N));

    // FFT transform via ND
    const UnsignedInteger rows = sample.getSize();
    const UnsignedInteger cols = sample.getDimension();
    ComplexCollection sampleFlat(rows * cols);
    for (UnsignedInteger j = 0; j < cols; ++j)
      for (UnsignedInteger i = 0; i < rows; ++i)
        sampleFlat[i + j * rows] = Complex(sample(i, j), 0.0);
    const ComplexCollection transformedSample(myFFT.transform(sampleFlat, {rows, cols}));

    // Inverse transformation
    const ComplexCollection inverseTransformedSample(myFFT.inverseTransform(transformedSample, {rows, cols}));
    for (UnsignedInteger i = 0; i < sampleFlat.getSize(); ++i)
      assert_almost_equal(inverseTransformedSample[i], sampleFlat[i]);

    // 3D ND case
    Tensor tensor(N, N, N);
    for(UnsignedInteger k = 0; k < N; ++k)
      for (UnsignedInteger j = 0; j < N; ++j)
        for (UnsignedInteger i = 0; i < N; ++i)
          tensor(i, j, k) = RandomGenerator::Generate();

    // FFT transform via ND
    ComplexCollection tensorFlat(N * N * N);
    for (UnsignedInteger k = 0; k < N; ++k)
      for (UnsignedInteger j = 0; j < N; ++j)
        for (UnsignedInteger i = 0; i < N; ++i)
          tensorFlat[i + j * N + k * N * N] = Complex(tensor(i, j, k), 0.0);
    const ComplexCollection transformedTensor(myFFT.transform(tensorFlat, {N, N, N}));

    // Inverse transformation
    const ComplexCollection inverseTransformedTensor(myFFT.inverseTransform(transformedTensor, {N, N, N}));
    for (UnsignedInteger i = 0; i < tensorFlat.getSize(); ++i)
      assert_almost_equal(inverseTransformedTensor[i], tensorFlat[i]);


  }
  catch (const TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
